rfc/rfc021-vp_token-grant-type.md

```diff
@@ -0,0 +1,215 @@
+8. The `exp` field MUST be present and contain a valid unix timestamp. It MUST be after the current time.
+9. The difference between the `exp` and `iat` fields MUST be equal or less than 5 seconds.
+10. The `jti` field MUST be present and contain a string that is unique for each access token request.
+
+### 4.3 JSON format requirements
+
+1. The assertion MUST be a valid JSON object according to §4.10 the Verifiable Credentials Data Model 1.1 [VC-DATA-MODEL].
+2. The proof of the VP MUST be a valid [JSONWebSignature2020] object. Additional requirements for the proof object:
+   1. The `challenge` field MUST be a string that is unique for each token request.
+   2. The `domain` field MUST be a DID under control of the Authorization Server.
+   3. The `verificationMethod` field MUST be a DID URL.
+   4. The id part of the DID in the `verificationMethod` field MUST match the `credentialSubject.id` field from all the Verifiable Credentials that are used to request the access token.
+   5. The `created` field MUST be present and contain a valid ISO8601 formatted date string. It MUST be before the current time.
+   6. The `expires` field MUST be present and contain a valid ISO8601 formatted date string. It MUST be after the current time.
+   7. The difference between the `expires` and `created` fields MUST be equal or less than 5 seconds.
+
+### 4.4 Preventing Token Replay
+
+The Authorization Server MUST reject any request that uses the unique value (`jti` or `challenge`) that has been used before.
+This approach has been chosen over the `nonce` field because there's no initial request to get a nonce from the Authorization Server.
+The Authorization Server MUST store the unique value for 10 seconds and MUST reject any request that uses a unique value that has been used before.
+The 10 seconds is based on the 5-second clock skew and the 5-second maximum difference between the expires and issued fields.
+
+### 4.5 Error Response
+
+If the Authorization Server determines that the VP is invalid, the Authorization Server MUST return an error response as defined in OAuth 2.0 [RFC6749].
+In addition to the error response defined in OAuth 2.0 [RFC6749], the Authorization Server MUST return a HTTP 400 (Bad Request) and use the following error codes when the VP is invalid:
+
+* `invalid_verifiable_presentation`: The VP is invalid. This error code is used when the signature is incorrect or when a required field is missing.
+* `invalid_presentation_submission`: The Presentation Submission is invalid. This error code is used when the Presentation Submission is not an answer to the Presentation Definition that corresponds with the requested scope.
+
+An `invalid_request` is returned for any submitted Verifiable Credentials that do not meet the requirements, when the Verifiable Credentials aren't corresponding to the Presentation Definition or when the Verifiable Credentials are expired, not trusted or invalid. 
```

Comment:
Kinda ambiguous with "invalid_verifiable_presentation"; if the contained credentials are invalid, it could also be seen as "invalid_verifiable_presentation"?

Also, I think there (ideally) should not be any other "trust" anchors than the one specified by the Presentation Definition. Otherwise, RP and AS still have to agree on a set of trust anchors outside of this specification.

Reply:
We currently decided on putting trust anchors in the PDs, that does not mean that this is mandatory through this RFC. 'Trust' is therefore still required, but undefined how it's solved.